lib/dry/validation/extensions/predicates_as_macros.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require 'dry/logic/predicates'
+require 'dry/schema/predicate_registry'
+require 'dry/validation/contract'
+
+module Dry
+  module Validation
+    # Predicate registry with additional needed methods.
+    class PredicateRegistry < Schema::PredicateRegistry
+      # List of predicates to be imported by `:predicates_as_macros`
+      # extension.
+      #
+      # @see Dry::Validation::Contract
+      WHITELIST = %i[
+        filled? gt? gteq? included_in? includes? inclusion? is? lt?
+        lteq? max_size? min_size? not_eql? odd? respond_to? size? true?
+        uuid_v4?
+      ].freeze
+
+      # @api private
+      def arg_names(name)
+        arg_list(name).map(&:first)
+      end
+
+      # @api private
+      def call(name, args)
+        self[name].(*args)
+      end
+
+      # @api private
+      def message_opts(name, arg_values)
+        arg_names(name).zip(arg_values).to_h
+      end
+    end
+
+    # Extension to use dry-logic predicates as macros.
+    #
+    # @see Dry::Validation::PredicateRegistry::WHITELIST Available
+    # predicates
+    #
+    # @example
+    #   Dry::Validation.load_extensions(:predicates_as_macros)
+    #
+    #   class ApplicationContract < Dry::Validation::Contract
+    #     import_predicates_as_macros
+    #   end
+    #
+    #   class AgeContract < ApplicationContract
+    #     schema do
+    #       required(:name).filled(:integer)
+    #     end
+    #
+    #     rule(:age).validate(gteq?: 18)
+    #   end
+    #
+    #   AgeContract.new.(age: 17).errors.first.text
+    #   # => 'must be greater than or equal to 18'
+    class Contract
+      # Make macros available for self and its descendants.
+      def self.import_predicates_as_macros
+        registry = PredicateRegistry.new
+        PredicateRegistry::WHITELIST.each do |name|
+          register_macro(name) do |macro:|
+            predicate_args = [*macro.args, value]
+            registry.(name, predicate_args) ||
+              key.failure(name, registry.message_opts(name, predicate_args))
```

Comment:
can't we just do:

```ruby
key.failure(name, registry.message_opts(name, predicate_args)) unless registry[name].(*predicate_args)
```

?

Reply:
I'm sometimes a little fuzzy and prefer using expressions with explicit return values (`if` without `else` returns `nil` on `false` branch, but, wait, is `nil` in any case a value???) :smile: 

Anyway, not a big deal, last commit contains this change. However, rubocop doesn't like it and I should use a guard clause, because a one-line modifier would be too long. Do I change it?

Reply:
it's easier to read and understand a conditional (at least for me) and ACTUALLY this should use `or` if we wanted to write more idiomatic Ruby because it's not about returning a value, it's about setting a failure but anyway, a conditional is clearer so let's stick to it.

> Do I change it?

No it's fine, I'll merge and do some quick follow-up work anyway